src/Gas.sol

```diff
@@ -1,332 +1,178 @@
+    } // Changed layout and type
 
-    event AddedToWhitelist(address userAddress, uint256 tier);
+    // State Variables
+    uint256 private immutable totalSupply; // Removed Initialization & changed to immutable
+    uint256 private paymentCounter = 0;
 
-    modifier onlyAdminOrOwner() {
-        address senderOfTx = msg.sender;
-        if (checkForAdmin(senderOfTx)) {
-            require(
-                checkForAdmin(senderOfTx),
-                "Gas Contract Only Admin Check-  Caller not admin"
-            );
-            _;
-        } else if (senderOfTx == contractOwner) {
-            _;
-        } else {
-            revert(
-                "Error in Gas contract - onlyAdminOrOwner modifier : revert happened because the originator of the transaction was not the admin, and furthermore he wasn't the owner of the contract, so he cannot run this function"
-            );
-        }
-    }
+    bool wasLastOdd = true;
+    address private contractOwner;
+    address[5] public administrators;
 
-    modifier checkIfWhiteListed(address sender) {
-        address senderOfTx = msg.sender;
-        require(
-            senderOfTx == sender,
-            "Gas Contract CheckIfWhiteListed modifier : revert happened because the originator of the transaction was not the sender"
-        );
-        uint256 usersTier = whitelist[senderOfTx];
-        require(
-            usersTier > 0,
-            "Gas Contract CheckIfWhiteListed modifier : revert happened because the user is not whitelisted"
-        );
-        require(
-            usersTier < 4,
-            "Gas Contract CheckIfWhiteListed modifier : revert happened because the user's tier is incorrect, it cannot be over 4 as the only tier we have are: 1, 2, 3; therfore 4 is an invalid tier for the whitlist of this contract. make sure whitlist tiers were set correctly"
-        );
-        _;
-    }
+    mapping(address => Payment[]) private payments;
+    mapping(address => uint256) public whitelist;
+    mapping(address => uint256) public balances;
+    mapping(address => bool) private isOddWhitelistUser;
+    mapping(address => uint256) private whiteListStruct;
 
+    // Events
+    event AddedToWhitelist(address userAddress, uint256 tier);
     event supplyChanged(address indexed, uint256 indexed);
-    event Transfer(address recipient, uint256 amount);
+    event Transfer(address indexed recipient, uint256 indexed amount);
     event PaymentUpdated(
-        address admin,
-        uint256 ID,
+        address indexed admin,
+        uint256 indexed ID,
         uint256 amount,
-        string recipient
+        bytes indexed recipient
     );
     event WhiteListTransfer(address indexed);
 
     constructor(address[] memory _admins, uint256 _totalSupply) {
         contractOwner = msg.sender;
         totalSupply = _totalSupply;
 
-        for (uint256 ii = 0; ii < administrators.length; ii++) {
+        for (uint256 ii = 0; ii < 5 /*administrators.length*/; ) {
```

Comment:
You can actually change the 5 to a constant. As far as I remember that saves gas if you use that constant twice.